security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,428 @@
+                        self._srk = srkPublicKey;
+                        self._fsm.transition('connected', callback);
+                      } else {
+                        self._fsm.transition('disconnected', callback, srkCreateErr);
+                      }
+                    });
+                  } else {
+                    self._fsm.transition('disconnected', callback, ekCreateErr);
+                  }
+                });
+              });
+            } catch (err) {
+              this._fsm.transition('disconnected', callback, err);
+            }
+          },
+          '*': () => this._fsm.deferUntilTransition()
+        },
+        connected: {
+          _onEnter: (callback, err) => {
+            callback(null);
+          },
+          getEndorsementKey: (callback) => {
+            callback(null, this._ek.asTpm2B());
+          },
+          getStorageRootKey: (callback) => {
+            callback(null, this._srk.asTpm2B());
+          },
+          signWithIdentity: (deviceIdData, callback) => {
+            this._signData(deviceIdData, (errSign: Error, signedData: Buffer) => {
+              if (errSign) {
+                debug('Error from signing data: ' + errSign);
+                callback(errSign);
+                this._fsm.transition('disconnected');
+              } else {
+                callback(null, signedData);
+              }
+            });
+          },
+          activateSymmetricIdentity: (identityKey, callback) => {
+            this._activateSymetricIdentity(identityKey, (errActivate: Error) => {
+              if (errActivate) {
+                debug('Error from activate: ' + errActivate);
+                callback(errActivate);
+                this._fsm.transition('disconnected');
+              } else {
+                callback(null, null);
+              }
+            });
+          },
+        }
+      }
+    });
+  }
+
+  getEndorsementKey(callback: (err: Error, endorsementKey: string) => void): void {
+      this._fsm.handle('getEndorsementKey', callback);
+  }
+
+  getStorageRootKey(callback: (err: Error, storageKey: string) => void): void {
+    this._fsm.handle('getStorageRootKey', callback);
+  }
+
+  signWithIdentity(deviceIdData: Buffer, callback: (err: Error, signedResponse: Buffer) => void): void {
+    if (deviceIdData === null || deviceIdData.length === 0) {
+        throw new ReferenceError('\'deviceIdData\' cannot be \'' + deviceIdData + '\'');
+    }
+
+    if (this._idKeyPub == null) {
+        throw new errors.ArgumentError('activateIdentityKey first before signing');
+    }
+
+    this._fsm.handle('signWithIdentity', deviceIdData, callback);
   }
 
-  getStorageRootKey(callback: (err?: Error) => void): void {
-    throw new errors.NotImplementedError();
+  activateSymmetricIdentity(identityKey: Buffer, callback: (err: Error, returnedActivate: Buffer) => void): void {
+    this._fsm.handle('activateSymmetricIdentiy', identityKey, callback);
```

Comment:
>handle [](start = 14, length = 6)

no argument validation for identityKey?

Reply:
considering the fact these might not be public APIs (in the sense that SDK users are not going to call them directly) i don't mind not having argument checking, but i do mind consistency across methods.

---
In reply to: [157889556](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r157889556) [](ancestors = 157889556)

Reply:
Fixed.

---
In reply to: [157889795](https://github.com/Azure/azure-iot-sdk-node/pull/185#discussion_r157889795) [](ancestors = 157889795,157889556)